net/core/filter.c

```diff
@@ -4792,6 +4792,255 @@ static const struct bpf_func_proto bpf_lwt_seg6_adjust_srh_proto = {
+	.arg1_type	= ARG_ANYTHING,
+	.arg2_type	= ARG_ANYTHING,
+	.arg3_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_pop, int, n)
+{
+	lua_pop(L, n);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pop_proto = {
+	.func		= bpf_lua_pop,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_pushinteger, lua_Integer, integer)
+{
+	lua_pushinteger(L, integer);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushinteger_proto = {
+	.func		= bpf_lua_pushinteger,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_pushlightuserdata, void *, ptr)
+{
+	lua_pushlightuserdata(L, ptr);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushlightuserdata_proto = {
+	.func		= bpf_lua_pushlightuserdata,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_2(bpf_lua_pushlstring, const char *, str, size_t, len)
+{
+	lua_pushlstring(L, str, len);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushlstring_proto = {
+	.func		= bpf_lua_pushlstring,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+	.arg2_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_pushmap, struct bpf_map *, map)
+{
+	lua_pushlightuserdata(L, map);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushmap_proto = {
+	.func		= bpf_lua_pushmap,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_CONST_MAP_PTR,
+};
+
+
+BPF_LUA_CALL_0(bpf_lua_pushskb)
+{
+	lua_pushlightuserdata(L, xdp_lua_get_skb());
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushskb_proto = {
+	.func		= bpf_lua_pushskb,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+};
+
+BPF_LUA_CALL_1(bpf_lua_pushstring, const char *, str)
+{
+	lua_pushstring(L, str);
+	return 0;
+}
+
+static const struct bpf_func_proto bpf_lua_pushstring_proto = {
+	.func		= bpf_lua_pushstring,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_toboolean, int, index)
+{
+	return lua_toboolean(L, index);
+}
+
+static const struct bpf_func_proto bpf_lua_toboolean_proto = {
+	.func		= bpf_lua_toboolean,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_1(bpf_lua_tointeger, int, index)
+{
+	return lua_tointeger(L, index);
+}
+
+static const struct bpf_func_proto bpf_lua_tointeger_proto = {
+	.func		= bpf_lua_tointeger,
+	.gpl_only	= false,
+	.pkt_access	= false,
+	.ret_type	= RET_INTEGER,
+	.arg1_type	= ARG_ANYTHING,
+};
+
+BPF_LUA_CALL_3(bpf_lua_tostring, char *, str, u32, size, int, index)
+{
+	if (!lua_isstring(L, index))
+		return 0;
+
+	strncpy(str, lua_tostring(L, index), size);
+	return 1;
```

Comment:
You could return the length of the string, probably more useful than 0 or -1.
Also, I think it is probably better to fail on bytes >= max/size, to to truncate and succeed.

```
BPF_LUA_CALL_3(bpf_lua_tostring, char *, o, u32, max, int, index)
{
	size_t bytes;
	const char *s = lua_tolstring(L, index, &bytes);

	if (!s || (bytes > max))
		return -1;
	strncpy(o, s, bytes);
	return bytes;
}
```